File: MathExpression.py
def get_precedence(op):
    if op in ['+', '-']:
        return 1
    elif op in ['*', '/']:
        return 2
    else:
        return 0

def perform_operation(num1, num2, op):
    if op == '+':
        return num1 + num2
    elif op == '-':
        return num1 - num2
    elif op == '*':
        return num1 * num2
    elif op == '/':
        return num1 / num2
    else:
        return 0

def evaluate_expression(expression):
    operands = []
    operators = []

    i = 0
    while i < len(expression):
        ch = expression[i]
        if ch == ' ':
            i += 1
            continue

        if ch.isdigit():
            num = 0
            while i < len(expression) and expression[i].isdigit():
                num = num * 10 + int(expression[i])
                i += 1
            operands.append(num)
        elif ch == '(':
            operators.append(ch)
            i += 1
        elif ch == ')':
            while operators and operators[-1] != '(':
                num2 = operands.pop()
                num1 = operands.pop()
                op = operators.pop()
                result = perform_operation(num1, num2, op)
                operands.append(result)
            operators.pop()
            i += 1
        else:
            while operators and get_precedence(ch) <= get_precedence(operators[-1]):
                num2 = operands.pop()
                num1 = operands.pop()
                op = operators.pop()
                result = perform_operation(num1, num2, op)
                operands.append(result)
            operators.append(ch)
            i += 1

    while operators:
        num2 = operands.pop()
        num1 = operands.pop()
        op = operators.pop()
        result = perform_operation(num1, num2, op)
        operands.append(result)

    return operands[0]

expression = "3*(1+4/2)+3*1"
result = evaluate_expression(expression)
print("Result:", result)

File: MathExpression.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

// Operator precedence map
var precedence = map[string]int{
	"+": 1,
	"-": 1,
	"*": 2,
	"/": 2,
}

// Stack data structure
type Stack []string

// Push adds an element to the top of the stack
func (s *Stack) Push(val string) {
	*s = append(*s, val)
}

// Pop removes and returns the top element of the stack
func (s *Stack) Pop() string {
	index := len(*s) - 1
	element := (*s)[index]
	*s = (*s)[:index]
	return element
}

// Perform operation based on the operator
func performOperation(operands *Stack, operator string) {
	num2, _ := strconv.Atoi(operands.Pop())
	num1, _ := strconv.Atoi(operands.Pop())

	switch operator {
	case "+":
		operands.Push(strconv.Itoa(num1 + num2))
	case "-":
		operands.Push(strconv.Itoa(num1 - num2))
	case "*":
		operands.Push(strconv.Itoa(num1 * num2))
	case "/":
		operands.Push(strconv.Itoa(num1 / num2))
	}
}

// Evaluate the mathematical expression
func evaluateExpression(expression string) int {
	operators := Stack{}
	operands := Stack{}

	tokens := strings.Split(expression, "")

	for _, token := range tokens {
		if token == " " {
			continue
		}

		if _, err := strconv.Atoi(token); err == nil {
			operands.Push(token)
		} else if token == "(" {
			operators.Push(token)
		} else if token == ")" {
			for len(operators) > 0 && operators[len(operators)-1] != "(" {
				performOperation(&operands, operators.Pop())
			}
			operators.Pop() // Discard "("
		} else {
			for len(operators) > 0 && precedence[token] <= precedence[operators[len(operators)-1]] {
				performOperation(&operands, operators.Pop())
			}
			operators.Push(token)
		}
	}

	for len(operators) > 0 {
		performOperation(&operands, operators.Pop())
	}

	result, _ := strconv.Atoi(operands.Pop())
	return result
}

func main() {
	expression := "3*(4/2+1)+3*1"
	result := evaluateExpression(expression)
	fmt.Println("Result:", result) // Output: 12
}
